cdap-data-fabric/src/main/java/co/cask/cdap/data2/registry/UsageRegistry.java

```diff
@@ -52,14 +54,22 @@
   private final Transactional<UsageDatasetIterable, UsageDataset> txnl;
 
   @Inject
-  public UsageRegistry(TransactionExecutorFactory txExecutorFactory, final DatasetFramework datasetFramework) {
+  public UsageRegistry(TransactionExecutorFactory txExecutorFactory,
+                       final DatasetProvider provider,
+                       final DatasetFramework framework) {
     txnl = Transactional.of(txExecutorFactory, new Supplier<UsageDatasetIterable>() {
       @Override
       public UsageDatasetIterable get() {
         try {
-          Object usageDataset = DatasetsUtil.getOrCreateDataset(datasetFramework, USAGE_INSTANCE_ID,
-                                                                UsageDataset.class.getSimpleName(),
-                                                                DatasetProperties.EMPTY, null, null);
+          Object usageDataset = provider.get(USAGE_INSTANCE_ID, null, null);
+          if (usageDataset == null) {
+            // use DatasetFramework only to create
+            DatasetsUtil.createIfNotExists(framework, USAGE_INSTANCE_ID,
```

Comment:
Only when the usage registry dataset doesn't exist yet, but otherwise the first provider.get() should not return null.